src/Storages/MergeTree/MergeTreeDataMergerMutator.cpp

```diff
@@ -355,27 +355,37 @@ bool MergeTreeDataMergerMutator::selectPartsToMerge(
 
     LOG_DEBUG(log, "Selected {} parts from {} to {}", parts.size(), parts.front()->name, parts.back()->name);
     future_part.assign(std::move(parts));
-    return true;
+    return SelectPartsDecision::SELECTED;
 }
 
-bool MergeTreeDataMergerMutator::selectAllPartsToMergeWithinPartition(
+SelectPartsDecision MergeTreeDataMergerMutator::selectAllPartsToMergeWithinPartition(
     FutureMergedMutatedPart & future_part,
     UInt64 & available_disk_space,
     const AllowedMergingPredicate & can_merge,
     const String & partition_id,
     bool final,
-    String * out_disable_reason)
+    const StorageMetadataPtr & metadata_snapshot,
+    String * out_disable_reason,
+    const Context * context)
```

Comment:
Maybe pass context by const reference? I don't see any reason to use pointer, but it will be more convinient.

Reply:
I had the reason. `selectAllPartsToMergeWithinPartition` function is used in `selectPartsToMerge` from `StorageMergeTree.cpp`, and `selectPartsToMerge` has several usage in different places. One of them is here:
https://github.com/ClickHouse/ClickHouse/blob/623dc2df7af1d0fbe1ba1d44cc9fcd1d1a87339b/src/Storages/StorageMergeTree.cpp#L940
And it's seemed that it's difficult to get the context here, so I decided to use pointer with nullptr as default.

Reply:
Hmm, I thought, maybe it would be better to pass the actual setting `optimize_skip_merged_partitions` with default as false instead of context. @CurtizJ what do you think?

Reply:
Hmm, I thought, maybe it would be better to pass the actual setting `optimize_skip_merged_partitions` with default as false instead of context. @CurtizJ what do you think?

Reply:
It also will be ok. I think it's more convenient.